adoc/chapters/programming_interface.adoc

```diff
@@ -15935,7 +15949,10 @@ The following member functions provide various queries for a <<kernel>>.
 backend get_backend() const noexcept;
 ----
 
-_Returns:_ The backend associated with this kernel.
+_Returns:_ The backend (if any) that underlies this object.
```

Comment:
I specified `backend::none` as "not a backend".  Here is the description from the section on the `backend` enumeration:

> The enumerator value `backend::none` is used to identify a SYCL object that does not correspond to any backend.

In addition, the section "The SYCL backend model" describes backends this way:

> A SYCL implementation may expose devices by layering on top of a heterogeneous API such as OpenCL or it may expose devices in some other way. [...] When a SYCL implementation layers on top of a heterogeneous API such as OpenCL, we refer to this as a SYCL backend.

Thus, when an implementation exposes devices without layering on top of a heterogeneous API, it is not using a backend.

Reply:
Would you happen to know the reason for such a choice for restricting backend to "heterogeneous API" ? 

 That a `TBB` or a `OpenMP` are not considered backends seems odd. It makes sense that they don't provide any interop, but I think the general usage of the word "backend" if more general.
 
 For example, in the SimSYCL paper, they use the term "AdaptiveCpp (OMP backend)" (https://dl.acm.org/doi/fullHtml/10.1145/3648115.3648136 table 2) who make total sense to me, but are not technically correct.
 
 In short, I don't understand the rationale behind this spec subtility. Every backend can decide if they provide or not some interopts for whatever SYCL function they want. 

PS: This also led to a question about an MPI Implementation. As far as I know, MPI is not a "heterogeneous API" (just an API :) ), but  I guess we still want to be able to give some interrupt handle.

Reply:
> That a TBB or a OpenMP are not considered backends seems odd. It makes sense that they don't provide any interop, but I think the general usage of the word "backend" if more general.

Yes. I'm opposed to tying the notion of a "backend" to the ability of providing interop. Those two have been separate concepts so far in SYCL.

AdaptiveCpp treats OpenMP just like any other backend. I see no reason to change the backend model for "native CPU" approaches.

Reply:
The term "heterogeneous API" is used in the SYCL specification even before this PR.  As far as I know, there is no industry standard definition for this term.  I was assuming that OMP would also be considered a "heterogeneous API".

I'm not opposed to using a different term if someone has a suggestion.  Is there some better collective term for things like OpenCL, OpenMP, CUDA, etc.?  Maybe "offload API"?

Reply:
I recommend disconnecting "heterogeneous API" from any "backend" consideration.   But maybe it is not the best place to discuss this. 

My main point is that we tied too much of the SYCL spec to the "heterogeneous API" of the backend.
For example:
>A function object that can execute on a [device](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#device) exposed by a [SYCL backend API](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#backend-api) is called a [SYCL kernel function](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sycl-kernel-function).

That means that for an OpenMP pure CPU or Intel TBB, a lambda used in Q.submit is not a kernel function. That also means that we can have "devices" that are not part of any "backend"?

OMP for offloading I I agree, is "heterogenous"; for CPU only OpenMP I don't see why it's "heterogeneous".  Same for MPI. Or for a `pthread` CPU implementation of SYCL. 

I think all this "SYCL is build on top of a backend API" is a relic of OpenCL; we should clarify that it's just implementation details. SYCL provides a "heterogenous API" but doesn't need to one be implemented. The spec should be clear about it.

And I think the easiest way is to define loosely what is a backend; it's whatever the implementation wants it to be, and not tie it to any concept of "heterogenous API"

But yeah, maybe a far more broad conversation; sorry for the side-discussion.